numpy/_core/src/multiarray/stringdtype/dtype.c

```diff
@@ -193,18 +217,18 @@ _eq_comparison(int scoerce, int ocoerce, PyObject *sna, PyObject *ona)
 static PyArray_StringDTypeObject *
 common_instance(PyArray_StringDTypeObject *dtype1, PyArray_StringDTypeObject *dtype2)
 {
-    int eq = _eq_comparison(dtype1->coerce, dtype2->coerce, dtype1->na_object,
-                            dtype2->na_object);
+    PyObject *out_na_object = NULL;
 
-    if (eq <= 0) {
-        PyErr_SetString(
-                PyExc_ValueError,
-                "Cannot find common instance for unequal dtype instances");
+    if (stringdtype_compatible_na(
+                dtype1->na_object, dtype2->na_object, &out_na_object) == -1) {
+        PyErr_Format(PyExc_TypeError,
```

Comment:
One of these I never knows, but isn't the old `ValueError` more appropriate than the new `TypeError`? After all, the type is OK, it is the value of `object_na` that is not.

(Also, I assume an already set error is just overridden? I haven't done much error stuff in C)

Reply:
It's a type error because it's a problem with the dtype for an operation. It would be a ValueError if it was a problem with the array values, not the dtype.